File: examples/csharp/src/dxbridge/CommandLine.cs
namespace DXBridgeExamples.DxBridge;

internal sealed class CommandLine
{
    private readonly Dictionary<string, string> _values = new(StringComparer.OrdinalIgnoreCase);
    private readonly HashSet<string> _flags = new(StringComparer.OrdinalIgnoreCase);

    public CommandLine(string[] args, IEnumerable<string> valueOptions, IEnumerable<string> flagOptions)
    {
        var valueSet = new HashSet<string>(valueOptions, StringComparer.OrdinalIgnoreCase);
        var flagSet = new HashSet<string>(flagOptions, StringComparer.OrdinalIgnoreCase);

        for (var i = 0; i < args.Length; i++)
        {
            var arg = args[i];
            if (arg is "--help" or "-h" or "/?")
            {
                HelpRequested = true;
                continue;
            }

            if (flagSet.Contains(arg))
            {
                _flags.Add(arg);
                continue;
            }

            if (valueSet.Contains(arg))
            {
                if (i + 1 >= args.Length)
                {
                    throw new ArgumentException($"Missing value for {arg}");
                }

                _values[arg] = args[++i];
                continue;
            }

            throw new ArgumentException($"Unknown argument: {arg}");
        }
    }

    public bool HelpRequested { get; }

    public bool HasFlag(string name)
    {
        return _flags.Contains(name);
    }

    public string? GetString(string name, string? defaultValue = null)
    {
        return _values.TryGetValue(name, out var value) ? value : defaultValue;
    }

    public int GetInt(string name, int defaultValue)
    {
        var value = GetString(name, null);
        if (value is null)
        {
            return defaultValue;
        }

        if (!int.TryParse(value, out var parsed))
        {
            throw new ArgumentException($"Invalid integer for {name}: {value}");
        }

        return parsed;
    }

    public float GetFloat(string name, float defaultValue)
    {
        var value = GetString(name, null);
        if (value is null)
        {
            return defaultValue;
        }

        if (!float.TryParse(value, out var parsed))
        {
            throw new ArgumentException($"Invalid float for {name}: {value}");
        }

        return parsed;
    }

    public float[] GetFloatList(string name, int expectedCount, float[] defaultValue)
    {
        var value = GetString(name, null);
        if (value is null)
        {
            return defaultValue;
        }

        var parts = value.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        if (parts.Length != expectedCount)
        {
            throw new ArgumentException($"Expected {expectedCount} comma-separated floats for {name}, got: {value}");
        }

        var result = new float[expectedCount];
        for (var i = 0; i < parts.Length; i++)
        {
            if (!float.TryParse(parts[i], out result[i]))
            {
                throw new ArgumentException($"Invalid float in {name}: {parts[i]}");
            }
        }

        return result;
    }
}

File: examples/java/src/dxbridge/CommandLine.java
package dxbridge;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public final class CommandLine {
    private final Map<String, String> values = new HashMap<String, String>();
    private final Set<String> flags = new HashSet<String>();
    private boolean helpRequested;

    public CommandLine(String[] args, Set<String> valueOptions, Set<String> flagOptions) {
        parse(args, valueOptions, flagOptions);
    }

    public static Set<String> options(String... names) {
        return new HashSet<String>(Arrays.asList(names));
    }

    public boolean isHelpRequested() {
        return helpRequested;
    }

    public boolean hasFlag(String name) {
        return flags.contains(name);
    }

    public String getString(String name, String defaultValue) {
        String value = values.get(name);
        return value == null ? defaultValue : value;
    }

    public int getInt(String name, int defaultValue) {
        String value = values.get(name);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("Invalid integer for " + name + ": " + value);
        }
    }

    public float getFloat(String name, float defaultValue) {
        String value = values.get(name);
        if (value == null) {
            return defaultValue;
        }
        try {
            return Float.parseFloat(value);
        } catch (NumberFormatException ex) {
            throw new IllegalArgumentException("Invalid float for " + name + ": " + value);
        }
    }

    public float[] getFloatList(String name, int expectedCount, float[] defaultValue) {
        String value = values.get(name);
        if (value == null) {
            return defaultValue;
        }

        String[] parts = value.split(",");
        if (parts.length != expectedCount) {
            throw new IllegalArgumentException(
                "Expected " + expectedCount + " comma-separated floats for " + name + ", got: " + value
            );
        }

        float[] parsed = new float[expectedCount];
        for (int i = 0; i < parts.length; i++) {
            try {
                parsed[i] = Float.parseFloat(parts[i].trim());
            } catch (NumberFormatException ex) {
                throw new IllegalArgumentException("Invalid float in " + name + ": " + parts[i]);
            }
        }
        return parsed;
    }

    private void parse(String[] args, Set<String> valueOptions, Set<String> flagOptions) {
        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            if ("--help".equals(arg) || "-h".equals(arg) || "/?".equals(arg)) {
                helpRequested = true;
                continue;
            }

            if (flagOptions.contains(arg)) {
                flags.add(arg);
                continue;
            }

            if (valueOptions.contains(arg)) {
                if (i + 1 >= args.length) {
                    throw new IllegalArgumentException("Missing value for " + arg);
                }
                values.put(arg, args[++i]);
                continue;
            }

            throw new IllegalArgumentException("Unknown argument: " + arg);
        }
    }
}
